event-schema/counter_manager.go

```diff
@@ -36,13 +54,41 @@ func getAllFrequencyCounters(schemaHash string) []*FrequencyCounter {
 	return frequencyCounters
 }
 
-func getFrequencyCounter(schemaHash, key string) *FrequencyCounter {
+func getFrequencyCounterBounded(schemaHash, key string, bound int) *FrequencyCounter {
 	schemaVersionCounters, ok := countersCache[schemaHash]
 	if !ok {
 		schemaVersionCounters = make(map[string]*FrequencyCounter)
 		countersCache[schemaHash] = schemaVersionCounters
 	}
 
+	diff := bound - len(schemaVersionCounters)
+	// bound reached, not allowed adding more values.
+	if diff == 0 {
+		// Just check and return value from the map
+		// no need to add anything to it.
+		return schemaVersionCounters[key]
+	}
+
+	// If we have exceeded the bound, we need to trim it
+	// to the new bound. This way whatever we have stored in memory
+	// gets purged which will be flushed back to the database on a schedule.
+	if diff < 0 {
+
+		toDelete := -1 * diff
+		for k := range schemaVersionCounters {
+			if toDelete > 0 {
+				delete(schemaVersionCounters, k)
+				toDelete--
+			} else {
+				break
+			}
+		}
```

Comment:
I feel a bit unexpected that a get method, has also a delete side-effect. 

Could that happen during `populateFrequencyCounters` or another method that runs periodically?

Reply:
This is a good point. Let me think about better structuring it so the function don't have unintended side-effects